Support hybrid boot mode on x86_64 AMI images and set the AMI boot mode on image registration

thozza opened the pull request:
Main changes introduced by this PR:

- `distro.ImageType` interface has been extended with a new `BootMode()` options, which makes it possible to determine for an image type, which boot modes it supports. In specific distro implementations, this is determined based on the `platform.Platform` associated with the specific image type.
- Internal AWS cloud API has been extended to allow specifying the AMI boot mode when registering the image.
- AWS upload target options have been extended to contain the boot mode to be used by worker when registering the AMI.
- Both, the Cloud API and Weldr API have been extended to set the desired image type boot mode (`distro.ImageType.BootMode() return value) as the desired AMI boot type in the AWS upload target options.
- The x86_64 AMI / EC2 images on all distributions have been modified to support hybrid boot mode. As a result, the RHEL-8 x86_64 AMI / EC2 base partition table changes accordingly (separate `/boot`, added `/boot/efi` partition). To make the partitioning unified across RHEL-9 and RHEL-8 and on all supported RHEL-8 architectures, the aarch64 RHEL-8 AMI / EC2 partition table slightly changed as well.
- Relevant test cases have been extended to test newly added functionality.

Side effect cleanups by this PR:

- `s390x` platform implementation now returns sane values from `GetZiplSupport()` method.
- `rhelX` distro implementations structures have been cleaned up from `legacy` and `bootType` members. In addition, the packages needed to boot the image are now added only by the associated platform, not implicitly to the image's base package set.
- The GCE image type has been made UEFI-only, which was the intention since the beginning.

This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
  - extended both, the Cloud API (`api.sh`) and Weldr API (`aws.sh`) test cases 
- [ ] adequate documentation informing people about the change such as
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`

Information regarding our GitLab pipeline (Schutzbot):

CI will not be ran automatically if WIP label is applied or the PR is in DRAFT state, instead only
a link will be provided to the pipeline which can then be triggered manually if desired. To run the
CI automatically either switch the PR to ready or apply WIP+test label.

Outside contributors need manual approval from one of the osbuild-composer maintainers.

Schutzbot will only be triggered if all Tests jobs in GitHub workflow succeed.
-->

achilleas-k reviewed (approved):
LGTM!

One small comment: Since the boot mode depends strictly on the platform configuration, would it make sense to put the `BootMode()` logic on the `Platform` struct instead?  We could expose it through the `ImageType` as well (as a little passthrough one-line method) so it would work the same as it does now, but the selection logic would be contained in the `BasePlatform`.

This could be done in a followup.

achilleas-k commented:
Merging since we're all green.  No need to delay this.  My comment can be handled in a followup.

thozza commented:
> LGTM!
> 
> One small comment: Since the boot mode depends strictly on the platform configuration, would it make sense to put the `BootMode()` logic on the `Platform` struct instead? We could expose it through the `ImageType` as well (as a little passthrough one-line method) so it would work the same as it does now, but the selection logic would be contained in the `BasePlatform`.
> 
> This could be done in a followup.

I was thinking about the same thing, but omitted it from the PR. Since you suggested it, I'll move it in a follow-up 😉 

Originally I also removed the `bootable` from `imageType` implementations and replaced it with use of `BootMode()`, however it turned out that some ostree images are not marked as bootable, since you need to deploy them first to boot them. This is also something to potentially rethink as part of your rework...

achilleas-k commented:
> Originally I also removed the `bootable` from `imageType` implementations and replaced it with use of `BootMode()`, however it turned out that some ostree images are not marked as bootable, since you need to deploy them first to boot them. This is also something to potentially rethink as part of your rework...

Somewhere in some branch I was working on for the rework I created `Unbootable` platform variants.  That's meant for containers and the tarball and the point is that it shouldn't have anything boot related and also no kernel.  We can also define variants for ostree commits, where they don't define any bootable parameters (uefi vs bios is undefined for example), but still should contain a kernel.

thozza commented:
> > Originally I also removed the `bootable` from `imageType` implementations and replaced it with use of `BootMode()`, however it turned out that some ostree images are not marked as bootable, since you need to deploy them first to boot them. This is also something to potentially rethink as part of your rework...
> 
> Somewhere in some branch I was working on for the rework I created `Unbootable` platform variants. That's meant for containers and the tarball and the point is that it shouldn't have anything boot related and also no kernel. We can also define variants for ostree commits, where they don't define any bootable parameters (uefi vs bios is undefined for example), but still should contain a kernel.

I agree, but in reality, we mark e.g. `edge-commit` and `edge-container` image types also as NOT bootable, which is half-way true in the context of unbootable platform. Because the image type itself is not bootable on its own, but once deployed it is bootable. So unbootable platform would not make sense in this context. I think that we overloaded the meaning of `bootable` in `ImageType` a bit. If we redefine what it really means, we could definitely replace it.

IMHO, any existing platform which returns empty string from `GetBIOSPlatform()` and `GetUEFIVendor()` and `false` from `GetZiplSupport()` should be already considered not bootable (which is what I used to determine if an image type is "bootable"...
